presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -221,38 +227,105 @@ public static HiveSplitSource bucketed(
                         if (preloadSplitsForGroupedExecution && !finished.get()) {
                             return allSplitsLoaded.transformAsync(ignored -> borrowBatchAsync(bucketNumber, maxSize, function), executor);
                         }
-                        return queueFor(bucketNumber).borrowBatchAsync(maxSize, function);
+                        return splitsFor(bucketNumber).borrowBatchAsync(maxSize, function);
                     }
 
                     @Override
                     public void finish()
                     {
                         if (finished.compareAndSet(false, true)) {
-                            queues.values().forEach(AsyncQueue::finish);
+                            buckets.values().forEach(BucketedSplits::finish);
                             allSplitsLoaded.set(null);
                         }
                     }
 
                     @Override
                     public boolean isFinished(OptionalInt bucketNumber)
                     {
-                        return queueFor(bucketNumber).isFinished();
+                        return splitsFor(bucketNumber).isFinished();
+                    }
+
+                    @Override
+                    public int rewind(OptionalInt bucketNumber)
+                    {
+                        checkState(allSplitsLoaded.isDone(), "rewind is only compatible when all splits are loaded");
+                        return splitsFor(bucketNumber).rewind();
                     }
 
-                    private AsyncQueue<InternalHiveSplit> queueFor(OptionalInt bucketNumber)
+                    private BucketedSplits splitsFor(OptionalInt bucketNumber)
                     {
                         checkArgument(bucketNumber.isPresent());
                         AtomicBoolean isNew = new AtomicBoolean();
-                        AsyncQueue<InternalHiveSplit> queue = queues.computeIfAbsent(bucketNumber.getAsInt(), ignored -> {
+
+                        BucketedSplits splits = buckets.computeIfAbsent(bucketNumber.getAsInt(), ignored -> {
                             isNew.set(true);
-                            return new AsyncQueue<>(estimatedOutstandingSplitsPerBucket, executor);
+                            return new BucketedSplits(estimatedOutstandingSplitsPerBucket, executor);
```

Comment:
The offer method never blocks. `estimatedOutstandingSplitsPerBucket` creates a false impression as if limits the queue size, when in fact it doesn't. I would remove this altogether.